### PR TITLE
Propagate request timeout to Nexus tasks

### PR DESCRIPTION
**Pull request by pdoerner**
## What changed?

Added handling to forward the `Request-Timeout` header supplied on user requests to workers.
Nexus clients set this header as a gRPC context deadline, so the value of the header is updated just before returning the task to subtract any processing time to that point.

## Why?

So that workers polling for Nexus tasks know how long they have to complete them.

## How did you test it?

Functional tests

### CODE REVIEW

File: common/nexus/constants.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nexus
+
+const HeaderRequestTimeout = "Request-Timeout"
```
Comment (bergundy) [inline common/nexus/constants.go:25]: Might as well export this from the SDK.

File: tests/nexus_api_test.go [hunk 2]
```diff
@@ -363,6 +423,8 @@ func (s *ClientFunctionalSuite) TestNexusStartOperation_Forbidden() {
 func (s *ClientFunctionalSuite) TestNexusStartOperation_Claims() {
 	taskQueue := s.randomizeStr("task-queue")
 	testService := s.createNexusIncomingService(s.randomizeStr("test-service"), taskQueue)
+	// Wait to make sure incoming service is loaded into memory before starting tests.
+	time.Sleep(3 * time.Second)
```
Comment (bergundy) [inline tests/nexus_api_test.go:427]: This is always a smell, you can use `EventuallyWithT` and assert that the service is ready.
Can do it in the next PR though.

File: tests/nexus_api_test.go [hunk 3]
```diff
@@ -511,23 +570,6 @@ func (s *ClientFunctionalSuite) TestNexusCancelOperation_Outcomes() {
 				s.Equal("deliberate internal failure", unexpectedError.Failure.Message)
 			},
 		},
-		// TODO: This can't be tested without the test taking over a minute since this is the default matching
```
Comment (bergundy) [inline tests/nexus_api_test.go:514]: Wondering why you didn't test this here.

File: tests/nexus_api_test.go [hunk 4]
```diff
@@ -270,9 +255,87 @@ func (s *ClientFunctionalSuite) TestNexusStartOperation_WithNamespaceAndTaskQueu
...
+		assertion       func(*nexus.ClientStartOperationResult[string], error)
+	}
+	testCases := []testcase{
+		{
+			name:            "header timeout > client timeout",
+			incomingService: s.createNexusIncomingService(s.randomizeStr("test-service"), s.randomizeStr("task-queue")),
+			headerTimeout:   3 * time.Second,
+			clientTimeout:   2 * time.Second,
+			assertion: func(res *nexus.ClientStartOperationResult[string], err error) {
+				s.ErrorIs(err, context.DeadlineExceeded)
```
Comment (bergundy) [inline tests/nexus_api_test.go:273]: This is basically testing the client, I don't think it deserves a test here.

File: tests/nexus_api_test.go [hunk 5]
```diff
@@ -575,6 +617,78 @@ func (s *ClientFunctionalSuite) TestNexusCancelOperation_Outcomes() {
 	}
 }
 
+func (s *ClientFunctionalSuite) TestNexusCancelOperation_Timeouts() {
```
Comment (bergundy) [inline tests/nexus_api_test.go:620]: There's no coverage for metrics tags here.

File: tests/nexus_api_test.go (hunk 3 above)
Comment (pdoerner) [reply tests/nexus_api_test.go:514]: Would have had to refactor the test cases for them to all set a header field which would have only been different for this test case. Seemed like a separate test method was easier

File: tests/nexus_api_test.go (hunk 4 above)
Comment (pdoerner) [reply tests/nexus_api_test.go:273]: Just wanted to be explicit about expectations on errors returned, but sure I can remove it

**Pull request closed without merging**